java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonSplitScanner.java

```diff
@@ -0,0 +1,214 @@
+        } catch (Catalog.TableNotExistException e) {
+            LOG.error("Failed to open the paimon reader.", e);
+            throw new IOException("Failed to open the paimon reader.", e);
+        }
+    }
+
+    private void parseRequiredTypes() {
+        List<String> fieldNames = PaimonScannerUtils.fieldNames(table.rowType());
+        requiredTypes = new ColumnType[requiredFields.length];
+        logicalTypes = new DataType[requiredFields.length];
+        for (int i = 0; i < requiredFields.length; i++) {
+            int index = fieldNames.indexOf(requiredFields[i]);
+            if (index == -1) {
+                throw new RuntimeException(String.format("Cannot find field %s in schema %s",
+                        requiredFields[i], fieldNames));
+            }
+            DataType dataType = table.rowType().getTypeAt(index);
+            String type = PaimonTypeUtils.fromPaimonType(dataType);
+            requiredTypes[i] = new ColumnType(type);
+            logicalTypes[i] = dataType;
+        }
+    }
+
+    private void initReader() throws IOException {
+
+        ReadBuilder readBuilder = table.newReadBuilder();
+        RowType rowType = table.rowType();
+        List<String> fieldNames = PaimonScannerUtils.fieldNames(rowType);
+        if (requiredFields.length < fieldNames.size()) {
+            int[] projected = Arrays.stream(requiredFields).mapToInt(fieldNames::indexOf).toArray();
+            readBuilder.withProjection(projected);
+        }
+        // TODO(wangriyu): readBuilder.withFilter()
+        // TODO(wangriyu): table.copy() tableWithDynamicOptions()
+        Split split = PaimonScannerUtils.decodeStringToObject(splitInfo);
+        reader = readBuilder.newRead().createReader(split);
+    }
+
+    @Override
+    public void open() throws IOException {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            initTable();
+            parseRequiredTypes();
+            initOffHeapTableWriter(requiredTypes, requiredFields, fetchSize);
+            initReader();
+        } catch (Exception e) {
+            close();
+            LOG.error("Failed to open the paimon reader.", e);
+            throw new IOException("Failed to open the paimon reader.", e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            if (reader != null) {
+                reader.close();
+            }
+        } catch (IOException e) {
+            LOG.error("Failed to close the paimon reader.", e);
+            throw new IOException("Failed to close the paimon reader.", e);
+        }
+    }
+
+    @Override
+    public int getNext() throws IOException {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            InternalRow row = null;
+            if (batch == null || (row = batch.next()) == null) {
```

Comment:
Does the `next()` method have a high overhead, will it be reused later?

Reply:
If the dataset size is larger than required, it will be reused